---
title: "AMR_deseq2"
author: "J. Meyer"
date: "2023-06-06"
output: html_document
---

Prepare data for DESeq2: 1. Count data should be formatted with genes in rows and samples in columns. 2. Metadata for the samples should be formatted with samples in rows and metadata in columns.


```{r, echo=FALSE}

counts <- read.table("20-samples-Bact-cds-counts.txt",sep="\t",header=TRUE)

# DeSeq is looking for only counts, without a first column of gene names, make the gene names row names.
mat <- counts[,-1]
rownames(mat) <- counts[,1]
# The Salmon counts have decimal places - round them up to prevent issues in deseq
counts2<-round(mat,digits=0)

# remove the disease samples from this analysis
counts2$AMR1D <- NULL
counts2$AMR2D <- NULL
counts2$AMR3D <- NULL
counts2$AMR5D <- NULL
counts2$AMR7D <- NULL
counts2$AMR8D <- NULL
counts2$AMR9D <- NULL

counts3 <-counts2[rowSums(counts2[])>0,] # remove empty rows
# use counts3 as input for DESeq2 
counts4 <- counts3
counts4$row_names <- row.names(counts4) # add back transcript.id as a column to merge with metadata
colnames(counts4)[14] <- 'transcript_id'
rgi <- read.table("rgi_out.txt",sep="\t",header=TRUE,quote="")
rgi_counts <-merge(rgi,counts4,by="transcript_id")
write.table(rgi_counts,"Untreated-vs-antibiotic_RGI-counts.txt",sep="\t",col.names=NA)

meta <-read.table("metadata.txt",sep="\t",header=TRUE)
# remove disease samples to compare just before and after antibiotic treatment
meta2 <-meta[meta$type != "disease",]
write.table(meta2,"metadata_13.txt",sep="\t",col.names=NA)

```

Summarize RGI genes by AMR Gene Family, Drug Class, and Resistance Mechanism in untreated vs antibiotic colonies; prepare data for DESeq2 analysis

```{r, echo=FALSE}
library(dplyr)
rgi_counts <- read.table("Untreated-vs-antibiotic_RGI-counts.txt",sep="\t",header=TRUE)
rgi_counts[1] <- NULL

#combine individual gene counts to get counts by AMR.Gene.Family
amrfam <- rgi_counts %>% group_by(AMR.Gene.Family) %>% summarize_at(vars(AMR1A:AMR10A), sum)
write.table(amrfam,"Untreated-vs-antibiotic_AMRgenefamily-counts.txt",sep="\t",col.names=NA)
#format for DESeq2
amrfam2 <- as.data.frame(amrfam)
rownames(amrfam2) <- amrfam2[,1]
amrfam2$AMR.Gene.Family <- NULL
# use amrfam2 as input for DESeq2

#combine individual gene counts to get counts by Drug.Class
drug <- rgi_counts %>% group_by(Drug.Class) %>% summarize_at(vars(AMR1A:AMR10A), sum)
write.table(drug,"Untreated-vs-antibiotic_DrugClass-counts.txt",sep="\t",col.names=NA)
#format for DESeq2
drug2 <- as.data.frame(drug)
rownames(drug2) <- drug2[,1]
drug2$Drug.Class <- NULL
# use drug2 as input for DESeq2

#combine individual gene counts to get counts by Resistance.Mechanism
mech <- rgi_counts %>% group_by(Resistance.Mechanism) %>% summarize_at(vars(AMR1A:AMR10A), sum)
write.table(mech,"Untreated-vs-antibiotic_ResistanceMechanism-counts.txt",sep="\t",col.names=NA)
#format for DESeq2
mech2 <- as.data.frame(mech)
rownames(mech2) <- mech2[,1]
mech2$Resistance.Mechanism <- NULL
# use mech2 as input for DESeq2

```

Run DESeq2 to test for differentially expressed *individual RGI genes* between coral colonies immediately before treatment ("untreated") and one day after amoxicillin treatment ("antibiotic").


```{r, echo=FALSE}
library(DESeq2)

meta <-read.table("metadata_13.txt",sep="\t",header=TRUE)
meta$X <- NULL
meta$type <-factor(meta$type)

# DESeq2 is looking for only counts, without a first column of gene names, make the gene names row names.
# This is testing all individual RGI genes
#construct DESEQDataSet Object
dds <- DESeqDataSetFromMatrix(countData=counts3, 
                              colData=meta, 
                              design=~type)
#see what the object looks like
dds

#now run DESeq function
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients
res <- results(dds, name="type_untreated_vs_antibiotic")
summary(res)

sink("DESeq2_summary_transcriptid.txt")
print(summary(res))
sink()

sink("DESeq2_results_transcriptid.txt")
print(res)
sink()

```


Run DESeq2 to test for differentially expressed *AMR gene families* between coral colonies immediately before treatment ("untreated") and one day after amoxicillin treatment ("antibiotic").


```{r, echo=FALSE}
library(DESeq2)

meta <-read.table("metadata_13.txt",sep="\t",header=TRUE)
meta$X <- NULL
meta$type <-factor(meta$type)

# DESeq2 is looking for only counts, without a first column of gene names, make the gene names row names.
# This is testing differential expression of AMR gene families
#construct DESEQDataSet Object
dds <- DESeqDataSetFromMatrix(countData=amrfam2, 
                              colData=meta, 
                              design=~type)
#see what the object looks like
dds

#now run DESeq function
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients
res <- results(dds, name="type_untreated_vs_antibiotic")
summary(res)

sink("DESeq2_results_AMRgenefamily.txt")
print(res)
sink()

sink("DESeq2_summary_AMRgenefamily.txt")
print(summary(res))
sink()
```

Run DESeq2 to test for differentially expressed *Drug Class* between coral colonies immediately before treatment ("untreated") and one day after amoxicillin treatment ("antibiotic").


```{r, echo=FALSE}
library(DESeq2)

meta <-read.table("metadata_13.txt",sep="\t",header=TRUE)
meta$X <- NULL
meta$type <-factor(meta$type)

# DESeq2 is looking for only counts, without a first column of gene names, make the gene names row names.
# This is testing differential expression of AMR gene families
#construct DESEQDataSet Object
dds <- DESeqDataSetFromMatrix(countData=drug2, 
                              colData=meta, 
                              design=~type)
#see what the object looks like
dds

#now run DESeq function
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients
res <- results(dds, name="type_untreated_vs_antibiotic")
summary(res)

sink("DESeq2_results_DrugClass.txt")
print(res)
sink()

sink("DESeq2_summary_DrugClass.txt")
print(summary(res))
sink()
```

Run DESeq2 to test for differentially expressed *Resistance Mechanism* between coral colonies immediately before treatment ("untreated") and one day after amoxicillin treatment ("antibiotic").


```{r, echo=FALSE}
library(DESeq2)

meta <-read.table("metadata_13.txt",sep="\t",header=TRUE)
meta$X <- NULL
meta$type <-factor(meta$type)

# DESeq2 is looking for only counts, without a first column of gene names, make the gene names row names.
# This is testing differential expression of AMR gene families
#construct DESEQDataSet Object
dds <- DESeqDataSetFromMatrix(countData=mech2, 
                              colData=meta, 
                              design=~type)
#see what the object looks like
dds

#now run DESeq function
dds <- DESeq(dds)
resultsNames(dds) # lists the coefficients
res <- results(dds, name="type_untreated_vs_antibiotic")
summary(res)

sink("DESeq2_results_ResistanceMechanism.txt")
print(res)
sink()

sink("DESeq2_summary_ResistanceMechanism.txt")
print(summary(res))
sink()
```


Since nothing is differentially expressed between untreated and antibiotic treated colonies, plot summary of the types of AMR genes that were expressed. I will use Resistance Mechanism since there are 11 types in this category. (Drug Class = 140 types, AMR gene families = 371 types)

```{r, echo=FALSE}
library(ggplot2)
library(reshape2)

meta <-read.table("metadata_13.txt",sep="\t",header=TRUE)
meta$X <- NULL
mech <-read.table("Untreated-vs-antibiotic_ResistanceMechanism-counts.txt",sep="\t",header=TRUE)
mech$X <- NULL
#calculate relative abundance from counts
mech_RA <- mech[,-1]/colSums(mech[,-1]) * 100
mech_RA$Resistance.Mechanism <- mech$Resistance.Mechanism

#merge metadata and count dataframes - start by converting mech from wide to long format
mech_long <- melt(mech_RA, id.vars=c("Resistance.Mechanism"))
colnames(mech_long)[colnames(mech_long) == "variable"] <- "sample"
colnames(mech_long)[colnames(mech_long) == "value"] <- "proportion"
resist <-merge(meta, mech_long, "sample")

#plot as stacked bars?

```

